### PR TITLE
Fix Bug 1470170 - Implement priority system for AS Router messages

### PR DESCRIPTION
**Pull request by sarracini**
This PR allows for a notion of "triggers" which act as a way to have priority in AS Router. This means that messages can be both targeted, for things that are long lasting (e.g locale, list of addons, prefs etc) but they can also be shown when a trigger happens - a one time event (e.g first new tab on a fresh profile, new addon installed, etc). Basically the content requests that a trigger message be shown under some condition (in this case we check the url to be about:welcome, indicating we are in a first run case) and then AS Router searches for messages that should be shown on a 'firstRun' trigger. Picks those, and shows them.

New method  `findMatchingMessageWithTrigger` called when content sends data 'trigger', and then fall back to the original `findMatchingMessage` if no messages with the trigger are found

### CODE REVIEW

File: lib/ASRouter.jsm [hunk 1]
```diff
@@ -251,7 +266,7 @@ class _ASRouter {
     } else {
       while (bundledMessagesOfSameTemplate.length) {
         // Find a message that matches the targeting context - or break if there are no matching messages
-        const message = await ASRouterTargeting.findMatchingMessage(bundledMessagesOfSameTemplate, target);
+        let message = await this._findMessage(bundledMessagesOfSameTemplate, target, data);
```
Comment (k88hudson) [inline lib/ASRouter.jsm]: why the change to `let`

Comment (sarracini) [reply lib/ASRouter.jsm]: because of copy pasting! i fix
